File: app/src/main/java/sfllhkhan95/doodle/core/views/ToolboxView.java
package sfllhkhan95.doodle.core.views;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;

import java.lang.reflect.Constructor;
import java.util.ArrayList;

import sfllhkhan95.doodle.R;
import sfllhkhan95.doodle.core.utils.OnToolSelectedListener;

/**
 * @author saifkhichi96
 */
public class ToolboxView extends LinearLayout {

    private final ArrayList<Integer> nonSticky = new ArrayList<>();

    private final LinearLayout primaryToolbox;
    private final LinearLayout secondaryToolbox;

    private int primarySelected = -1;
    private OnToolSelectedListener toolSelectedListener;

    public ToolboxView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        LayoutParams params = new LayoutParams(context, attrs);
        params.weight = 1;

        primaryToolbox = new LinearLayout(context, attrs);
        primaryToolbox.setLayoutParams(params);

        secondaryToolbox = new LinearLayout(context, attrs);
        secondaryToolbox.setLayoutParams(params);

        LayoutParams p = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
        setLayoutParams(p);

        this.setOrientation(VERTICAL);
        this.setWeightSum(2);

        addView(secondaryToolbox);
        addView(primaryToolbox);
        init();

        try {
            this.setOnToolSelectedListener((OnToolSelectedListener) context);
        } catch (Exception ignored) {

        }

        secondaryToolbox.setVisibility(GONE);
    }

    public ToolboxView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutParams params = new LayoutParams(context, attrs);
        params.weight = 1;

        primaryToolbox = new LinearLayout(context, attrs, defStyleAttr);
        primaryToolbox.setLayoutParams(params);

        secondaryToolbox = new LinearLayout(context, attrs, defStyleAttr);
        secondaryToolbox.setLayoutParams(params);

        LayoutParams p = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
        setLayoutParams(p);

        this.setOrientation(VERTICAL);
        this.setWeightSum(2);

        addView(secondaryToolbox);
        addView(primaryToolbox);
        init();

        try {
            this.setOnToolSelectedListener((OnToolSelectedListener) context);
        } catch (Exception ignored) {

        }

        secondaryToolbox.setVisibility(GONE);
    }

    private void inflateItem(final LinearLayout root, final MenuItem item) {
        LayoutParams params = new LayoutParams(0, LayoutParams.MATCH_PARENT);
        params.weight = 1;

        if (item.isCheckable()) {
            nonSticky.add(item.getItemId());
        }

        LinearLayout toolView = (LinearLayout) View.inflate(getContext(), R.layout.view_tool_icon, null);
        toolView.setLayoutParams(params);

        final ImageButton itemView = toolView.findViewById(R.id.icon);
        itemView.setId(item.getItemId());
        itemView.setImageDrawable(item.getIcon());

        itemView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                for (int i = 0; i < root.getChildCount(); i++) {
                    if (v.getId() == ((LinearLayout) root.getChildAt(i)).getChildAt(0).getId()) {
                        if (item.hasSubMenu()) {
                            if (primarySelected != v.getId() || secondaryToolbox.getVisibility() == GONE) {
                                inflateMenu(secondaryToolbox, item.getSubMenu());
                                secondaryToolbox.setVisibility(VISIBLE);

                                selectTool(primaryToolbox, v.getId());

                                if (hasOnToolSelectedListener()) {
                                    int id = ((LinearLayout) secondaryToolbox.getChildAt(0)).getChildAt(0).getId();
                                    selectTool(secondaryToolbox, id);

                                    toolSelectedListener.onToolSelected(!nonSticky.contains(id), id);
                                }
                            } else {
                                secondaryToolbox.setVisibility(GONE);

                                if (hasOnToolSelectedListener()) {
                                    toolSelectedListener.onToolSelected(!nonSticky.contains(v.getId()), v.getId());
                                }
                            }
                        } else {
                            if (root.equals(primaryToolbox)) {
                                secondaryToolbox.setVisibility(GONE);
                                selectTool(primaryToolbox, v.getId());
                            } else {
                                selectTool(secondaryToolbox, v.getId());
                            }

                            if (hasOnToolSelectedListener()) {
                                toolSelectedListener.onToolSelected(!nonSticky.contains(v.getId()), v.getId());
                            }
                        }
                    }
                }
            }
        });

        root.addView(toolView);
    }

    private boolean hasOnToolSelectedListener() {
        return toolSelectedListener != null;
    }

    private void inflateMenu(LinearLayout root, Menu menu) {
        root.removeAllViews();

        setWeightSum(menu.size());
        for (int i = 0; i < menu.size(); i++)
            inflateItem(root, menu.getItem(i));
    }

    private void init() {
        try {
            Class<?> menuBuilderClass = Class.forName("com.android.internal.view.menu.MenuBuilder");
            Constructor<?> constructor = menuBuilderClass.getDeclaredConstructor(Context.class);
            Menu menu = (Menu) constructor.newInstance(getContext());

            MenuInflater inflater = new MenuInflater(getContext());
            inflater.inflate(R.menu.tools, menu);

            inflateMenu(primaryToolbox, menu);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void setToolColor(int itemId, int color) {
        try {
            ((ImageButton) findViewById(itemId)).setColorFilter(color);
        } catch (Exception ignored) {

        }
    }

    public void selectTool(LinearLayout root, int id) {
        if (nonSticky.contains(id)) return;

        deselectAll(root);
        if (root.equals(primaryToolbox))
            primarySelected = id;

        setToolColor(id, Color.parseColor("#883997"));
    }

    public void deselectAll(LinearLayout root) {
        if (root.equals(primaryToolbox))
            primarySelected = -1;

        for (int i = 0; i < root.getChildCount(); i++) {
            setToolColor(((LinearLayout) root.getChildAt(i)).getChildAt(0).getId(), Color.WHITE);
        }
    }

    public void updateFillColorPicker(int color) {
        setToolColor(R.id.fillColorPicker, color);
    }

    public void updatePenColorPicker(int color) {
        setToolColor(R.id.penColorPicker, color);
    }

    public void setOnToolSelectedListener(OnToolSelectedListener toolSelectedListener) {
        this.toolSelectedListener = toolSelectedListener;
    }

}

File: app/src/main/java/sfllhkhan95/doodle/core/views/ToolboxView.kt
package sfllhkhan95.doodle.core.views

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.*
import android.widget.ImageButton
import android.widget.LinearLayout
import sfllhkhan95.doodle.R
import sfllhkhan95.doodle.core.utils.OnToolSelectedListener
import java.util.*

/**
 * @author saifkhichi96
 */
class ToolboxView : LinearLayout {

    private val nonSticky = ArrayList<Int>()

    private val primaryToolbox: LinearLayout
    private val secondaryToolbox: LinearLayout

    private var primarySelected = -1
    private var toolSelectedListener: OnToolSelectedListener? = null

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        val params = LinearLayout.LayoutParams(context, attrs)
        params.weight = 1f

        primaryToolbox = LinearLayout(context, attrs)
        primaryToolbox.layoutParams = params

        secondaryToolbox = LinearLayout(context, attrs)
        secondaryToolbox.layoutParams = params

        val p = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT)
        layoutParams = p

        this.orientation = LinearLayout.VERTICAL
        this.weightSum = 2f

        addView(secondaryToolbox)
        addView(primaryToolbox)
        init()

        try {
            this.setOnToolSelectedListener(context as OnToolSelectedListener)
        } catch (ignored: Exception) {

        }

        secondaryToolbox.visibility = View.GONE
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        val params = LinearLayout.LayoutParams(context, attrs)
        params.weight = 1f

        primaryToolbox = LinearLayout(context, attrs, defStyleAttr)
        primaryToolbox.layoutParams = params

        secondaryToolbox = LinearLayout(context, attrs, defStyleAttr)
        secondaryToolbox.layoutParams = params

        val p = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT)
        layoutParams = p

        this.orientation = LinearLayout.VERTICAL
        this.weightSum = 2f

        addView(secondaryToolbox)
        addView(primaryToolbox)
        init()

        try {
            this.setOnToolSelectedListener(context as OnToolSelectedListener)
        } catch (ignored: Exception) {

        }

        secondaryToolbox.visibility = View.GONE
    }

    private fun inflateItem(root: LinearLayout, item: MenuItem) {
        val params = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT)
        params.weight = 1f

        if (item.isCheckable) {
            nonSticky.add(item.itemId)
        }

        val toolView = View.inflate(context, R.layout.view_tool_icon, null) as LinearLayout
        toolView.layoutParams = params

        val itemView = toolView.findViewById<ImageButton>(R.id.icon)
        itemView.id = item.itemId
        itemView.setImageDrawable(item.icon)

        itemView.setOnClickListener { v ->
            for (i in 0 until root.childCount) {
                if (v.id == (root.getChildAt(i) as LinearLayout).getChildAt(0).id) {
                    if (item.hasSubMenu()) {
                        if (primarySelected != v.id || secondaryToolbox.visibility == View.GONE) {
                            inflateMenu(secondaryToolbox, item.subMenu)
                            secondaryToolbox.visibility = View.VISIBLE

                            selectTool(primaryToolbox, v.id)

                            if (hasOnToolSelectedListener()) {
                                val id = (secondaryToolbox.getChildAt(0) as LinearLayout).getChildAt(0).id
                                selectTool(secondaryToolbox, id)

                                toolSelectedListener!!.onToolSelected(!nonSticky.contains(id), id)
                            }
                        } else {
                            secondaryToolbox.visibility = View.GONE

                            if (hasOnToolSelectedListener()) {
                                toolSelectedListener!!.onToolSelected(!nonSticky.contains(v.id), v.id)
                            }
                        }
                    } else {
                        if (root == primaryToolbox) {
                            secondaryToolbox.visibility = View.GONE
                            selectTool(primaryToolbox, v.id)
                        } else {
                            selectTool(secondaryToolbox, v.id)
                        }

                        if (hasOnToolSelectedListener()) {
                            toolSelectedListener!!.onToolSelected(!nonSticky.contains(v.id), v.id)
                        }
                    }
                }
            }
        }

        root.addView(toolView)
    }

    private fun hasOnToolSelectedListener(): Boolean {
        return toolSelectedListener != null
    }

    private fun inflateMenu(root: LinearLayout, menu: Menu) {
        root.removeAllViews()

        weightSum = menu.size().toFloat()
        for (i in 0 until menu.size())
            inflateItem(root, menu.getItem(i))
    }

    private fun init() {
        try {
            val menuBuilderClass = Class.forName("com.android.internal.view.menu.MenuBuilder")
            val constructor = menuBuilderClass.getDeclaredConstructor(Context::class.java)
            val menu = constructor.newInstance(context) as Menu

            val inflater = MenuInflater(context)
            inflater.inflate(R.menu.tools, menu)

            inflateMenu(primaryToolbox, menu)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun setToolColor(itemId: Int, color: Int) {
        try {
            (findViewById<View>(itemId) as ImageButton).setColorFilter(color)
        } catch (ignored: Exception) {

        }

    }

    private fun selectTool(root: LinearLayout, id: Int) {
        if (nonSticky.contains(id)) return

        deselectAll(root)
        if (root == primaryToolbox)
            primarySelected = id

        setToolColor(id, Color.parseColor("#883997"))
    }

    private fun deselectAll(root: LinearLayout) {
        if (root == primaryToolbox)
            primarySelected = -1

        for (i in 0 until root.childCount) {
            setToolColor((root.getChildAt(i) as LinearLayout).getChildAt(0).id, Color.WHITE)
        }
    }

    fun updateFillColorPicker(color: Int) {
        setToolColor(R.id.fillColorPicker, color)
    }

    fun updatePenColorPicker(color: Int) {
        setToolColor(R.id.penColorPicker, color)
    }

    private fun setOnToolSelectedListener(toolSelectedListener: OnToolSelectedListener) {
        this.toolSelectedListener = toolSelectedListener
    }

}
